[NEW] Heavy weight threading - with minor model impact

mrsimpson opened the pull request:
<!-- INSTRUCTION: Inform the issue number that this PR closes, or remove the line below -->
Addresses #1112 and others

# What it does

This PR adds a simple form of threading to Rocket.Chat.

![threads-plain-rocketchat mp4](https://user-images.githubusercontent.com/17176678/44223688-99a2af00-a188-11e8-9c1b-66b0f3208310.gif)

# How it's done

Since there has been a lot of discussion about how threads should be implemented in RocketChat, I decided to take a first stab at it which has got the following properties:

- Minimal changes to the core of RC, to its files as well as the models, publications, subscriptions
- Configurable in such a way that users not liking it can disable it
- Compatible with the mobile apps
- Open to further enhancements

Looking at these non-functional requirements, I came up with the following:

- Each thread is a room of an existing room type (`c` or `p`) => this allows to lifecycle the rooms as usual, e. g. change their privacy, add/remove users and so on
- The rooms are linked only using the `parentRoomId`. No other change is applied to the models than adding this property to rooms and subscriptions (for performance)
- There is no inline-display of the thread in the parent room. Although this was possible with a good performance with this model, this involves significant modifications of the messages list - and it raises the question how to handle notifications.
- There is no harmonization of the create-room-modal and the thread-creation. This has the downside of bringing in an unwanted button in the sidebar but is less modification prone.

I believe, that with only low effort, we can really provide a configurable way of threading (per server and per user) which will satisfy most divergent demands.

I'm looking forward to your feedback - and contributions, if you want (that is directed to you, @vynmera  ;) )

Oliver

vynmera commented:
This looks very very good!!! I'll check it out locally asap :)

One UX issue I can imagine is the display of threads in the sidebar - this current display might cause confusion when you've got many threads (ops-1, ops-2, ops-3, ops-4, can you tell which is which?), so it'd be nice to have some sort of stub of the OP question in the thread's sidebar display or room name.

Perhaps we can even hide the parent room's name from the displayed name, and instead implement a tree-style structure (where threads are displayed below their parent channel with some extra margin-left), although that should possibly be toggleable per-user.

Also, do we have a permission to create threads? Perhaps some want to limit thread creation to specific trusted roles.

vynmera commented:
@mrsimpson Good to hear about the authorizations. 

As for the sidebar, that's totally fair, I can already see how the sidebar is going to be too thin for this (and it'll just say "How do I d..." a lot of times :sweat_smile:).

I'll take a shot at the message display when I have the time (I'll be back from vacation this Monday)!

engelgabriel commented:
This is an impressive start, and I am looking forward to adding this!

We have also been playing with the idea, and a design team created the following mockups:
https://projects.invisionapp.com/share/QVFYZPN26BH#/screens/280660916

mrsimpson commented on .circleci/config.yml:
```diff
@@ -285,27 +285,27 @@ jobs:
 
               echo "Build official Docker image"
               cp ~/repo/.docker/Dockerfile .
```
I'm aware that there are some files which should not be contributed but somehow resulted from our fork. I'll remove them once we finalized the requirement on our fork.

mrsimpson commented:
@geekgonecrazy 

I agree with the new-button not being nice where I put it. But on the other hand, I even didn't notice the button in the menu next to the message text input. I [extensively commented](https://projects.invisionapp.com/share/QVFYZPN26BH#/screens/280660916/comments) this and the other screens. We wanted to have maximum adoption of threading, so we decided to make it as prominent as possible.

With respect to the relationship of thread and parent: Any visualization of it has a major impact on the core product. Since I didn't have mockups yet, I decided to keep it as disruptiveless as possible. 
At the same time, I asked myself: "what does make it really necessary to have the link?" And seriously, except the members being similar, I don't see a need. Can you help me why this would screw your brain?

geekgonecrazy commented:
Its mostly just the case of a thread showing up as a private group that would ever really mess with me from a UX perspective.

@RocketChat/design definitely need thoughts here.  I think the Designed linked are old.  Threading before the whole new UI design was implemented.  Would be good to know which aspects of those designs would still fit in.

mrsimpson commented:
I updated to the latest `develop`, cleaned unnecessary files from our fork's merge and changed the thread-message-display: Now it looks quite similar compared to the mockups and I'm content with how it feels.
![image](https://user-images.githubusercontent.com/17176678/44752347-0324a500-ab1b-11e8-840c-994e48612cbf.png)

Biggest flaw compared to the mockups is - of course - the hierarchical view. But since I'm not a frontender I definitely need some help here. And it would be good to get some feedback from @RocketChat/design

zdumitru commented:
Is there any progress on this? Any blockers? Most requested feature ATM. Thanks for implementing it.

MAFLO321 commented:
@engelgabriel, @RocketChat/design and @RocketChat/core
Is there anything new about that?
As you might noticed in #1112, the thread feature is something what the users/community would really love since multiple years. Moreover, for some potential new user, this feature is a blocker to change from another platform to Rocket.Chat.

yevgenydevine commented:
Update this.

geekgonecrazy commented:
@MAFLO321 its definitely on our roadmap.  I think we are still trying to get UX figured out.  As you can see from that issue you linked to.. its a big topic and there are a looooot of opinions.  Almost no two are the same.  So this is a delicate thing that we would like to get right.  User experience is important to us.  

Also one has to remember mobile, what ever experience we decide on is going to have to be implemented on mobile.  So a lot of things to consider.  I know the team has on the backlog and will get to it as soon as they can :)

zdumitru commented:
Milestone here is set to 0.69.0, which is a version from the past. Does it need to be updated?

Ambyjkl commented:
@mrsimpson are you still actively working on this? Have you made any progress. This feature is really important and not having it is blocking my team from moving to rocketchat. I can help finish up the remaining stuff if you don't mind.

mrsimpson commented:
@Ambyjkl I'm currently not making progress here, since I'm waiting for the @RocketChat/core to provide concrete feedback and / or suggestions what's necessary for this PR to make it into `master`.
However, on our own fork, I have done some small enhancements (like who is selected as member for the thread and a `/thread`-command).
If this is blocking your team from moving to RC, feel free to use our fork. We're currently based on 0.70 (to be merged into `master` today) and we'll be soon updating to 0.71.1 (we're following a latest-1 strategy ;))

Ambyjkl commented:
@mrsimpson ok cool sounds good. Using your fork seems like a fine idea to me, I think we might give that a try. And thanks for implementing the feature!

zdumitru commented:
@geekgonecrazy Can this PR be reviewed, please?

geekgonecrazy commented:
@zdumitru as mentioned above we need UX and design involved since this is such an important feature.

mrsimpson commented:
@thiagosanchz 👍 ! I also thought about changes like these, but didn’t implement them yet since they need a lot of modification to existing templates and I wanted to keep it as modification free as possible. Once this is signed of in a „once we do this, we can merge it“ manner, I’ll be happy to update the PR, though I might need some small help. 

With respect to the changes: I think you have to be clear whether this threading replaces normal replies (I do think this was appropriate). Having both is kind-of confusing. Maybe, you could switch between the modes via configuration. 
Also, we‘d have to figure out the technical name of the thread (for the URI). I like the display name you chose. 
What do you suggest for the properties of the room? Is the initial question the description of the thread?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

pshute commented:
@mrsimpson I should have subscribed to this when you created it. I didn't realise you'd got so far.

Can you please explain a few thing?
- If there are lots of threads, say 1000, how are they displayed in the sidebar?
- If someone replies to a very old thread, what would draw anyone's attention to that reply?
- Have you done any work with displaying the thread replies in the parent channel? I'm very interested in the very "light threading" Flowdock style of display, where they're displayed chronologically, but coloured according to which thread they're in. Is this type of display possible? It would be like merging channels.

MarcosSpessatto commented on packages/assistify-threading/lib/threadRoomType.js:
```diff
@@ -0,0 +1,19 @@
+	}
+
+	condition() {
+		return RocketChat.getUserPreference(Meteor.userId(), 'sidebarShowThreads');
```
Should we not check the user's permissions here? :thinking:  Because the "create thread button" keeps showing up even with the permission disabled.

mrsimpson replied:
Yes, we should!

mrsimpson replied:
Added validation on authorization - but in another file ;)

mrsimpson commented:
@pshute sorry for the long delay, I simple did not get to care about this yet. I can only comment with respect to the implementation I have done. I don't claim it's perfect nor do I necessary think it should be this way, so please consider the following descriptive.

> * If there are lots of threads, say 1000, how are they displayed in the sidebar?

Each thread is a normal room. Each *subscribed* thread will be displayed as one item in the nav bar. Either in a separate section "Threads" or merged with the other channels, depending on the user's preference.

> * If someone replies to a very old thread, what would draw anyone's attention to that reply?

Just like with normal rooms: Once you've hidden a thread-channel, a new reply would re-pop into existence in the nav bar.

> * Have you done any work with displaying the thread replies in the parent channel? I'm very interested in the very "light threading" Flowdock style of display, where they're displayed chronologically, but coloured according to which thread they're in. Is this type of display possible with your system of threading? It would be like merging channels.

Model-wise, this is possible. However, the implementation which I've done is very heavy, meaning that the thread will not be inlined. I can very well imagine that there will be a display option in future. However, this is way more tricky than it seems at first glance. Some samples:

- How should old threads with new replies be sorted compared to messages which were written in the meantime?
- How is the user being notified about new items in multiple threads of the same channel?
-  How can one conveniently reply to a thread while also being able to ignore a thread?
- How would it match a mobile experience?

Don't get me wrong, there are answers to all of those questions. Each one has got tradeoffs which the maintainers need to decide upon. I can't. Therefore, I decided to got for an approach which is very heavy, but easily compatible also with mobile UX.

Hope this clarified your questions!

Cheers, 
Oliver

P.s.: In case you've got the urge to make implementation proposals for a more lightweight threading (addressing the questions above), please reply to #1112 so that those don't get lost if this PR focusing on the current style of implementation is closed

mrsimpson commented on packages/rocketchat-ui/client/components/header/header.js:
```diff
@@ -1,4 +1,4 @@
-/* globals fireGlobalEvent*/
+/* globals fireGlobalEvent, openRoom*/
```
I don't like to use this function, but I've not found another public / exported library function which navigates to a room

kclarkson commented:
@mrsimpson thank you so much for all of your contribution to this project. I am coming from the Drupal Community where we are slowly migrating from slack. The threading option is by far the most important feature that needs to be addressed. 

Your work will have an impact on hundreds of thousands of people!  Thank you 🥇

mrsimpson commented:
Todos based upon @thiagosanchz 's [Mockup above](https://github.com/RocketChat/Rocket.Chat/pull/11803#issuecomment-445110477) (WIP)

- [X] Add header with navigation
- [ ] Use message counter of thread (`... replies`) as navigational button on the message which started the thread
- [ ] Make the reply-counter reactive (separate task, since this might be tricky in the NRR-message-list)
- [ ] Move the thread-creation-button from the action list to next to the reactions-button (minor, imho)

To be discussed
- [ ] When navigating back, should we navigate to the message from which the thread started (currently only the room)
- [ ] (How) Should the original message be changed / denoted as threading-start? In my earlier implementation, I added a small prefix (`... started a thread`) quoting the original message. I believe this has got better UX compared to "just add the reply-counter" in order to signal it's a thread
- [ ] How does it fit in with the normal reply-action? Based upon @tassoevan 's suggestion, we could hijack the reply-action ;)

pshute commented:
@mrsimpson I'll reply here about these for now. I'll add comments in #1112 later once your work is released and others there understand what I'm talking about.

What I'm interested in is whether your scheme precludes the style of display I'm proposing, or makes it difficult to implement.

> How should old threads with new replies be sorted compared to messages which were written in the meantime?
In the Flowdock style of light threading, all messages are displayed chronologically in the original channel. The original channel and its thread channels would be displayed as merged into the original. The only indication of threading is that messages in the same thread are flagged with a unique colour.

> How is the user being notified about new items in multiple threads of the same channel?
There would be no special notifications for thread replies.

> How can one conveniently reply to a thread while also being able to ignore a thread?
With this style of display, one couldn't ignore a particular thread by hiding it. The idea of the colouring is that the user can see everything, but can physically look for red messages or ignore yellow ones.

> How would it match a mobile experience?
Because the only difference in the display would be the colours, there should be no problems with mobile displays.

geekgonecrazy commented:
This style would basically fork the conversation of a thread into its own room type.

So the flowdock style wouldn’t really be possible.  I imagine flowdock tags each message with roomid + threadid.  Then shows color based on thread.  Then when showing them together probably queries based on both id’s to return just the thread.

This style would be just like a normal room so if you wanted to you could tune in to the thread you actually want to participate in while leaving the rest of the channel alone.

You could then if you wanted post back to channel(not in scope of this PR but technically would be very easy to submit message quoting threads message to parent channel if we wanted)

> When navigating back, should we navigate to the message from which the thread started (currently only the room)

From mock-up seems like just navigate history back.  I think this is most helpful if going from channel to thread and then wanting to back out to channel again.

> (How) Should the original message be changed / denoted as threading-start? In my earlier implementation, I added a small prefix (`... started a thread`) quoting the original message. I believe this has got better UX compared to "just add the reply-counter" in order to signal it's a thread

I think original message should mostly be left alone.  If anything adding notation about reply right below makes it clear it’s a thread.  Maybe we could add an icon or something to make more clear there is a thread associated with that message now.

> How does it fit in with the normal reply-action? Based upon @tassoevan 's suggestion, we could hijack the reply-action ;)

I think should completely jack reply.  Unless it’s a DM... in a DM I think behavior should stay the same.  But that’s because I don’t think threads belong in a 1:1 conversation

pshute commented:
@geekgonecrazy 

> So the flowdock style wouldn’t really be possible. I imagine flowdock tags each message with roomid + threadid. Then shows color based on thread. Then when showing them together probably queries based on both id’s to return just the thread.

Is there no easy way to identify the original channel of a thread message with the system being worked on here?

mrsimpson commented:
>@geekgonecrazy  ...

@pshute as Aaron said, it’s kind of a different scheme. Given a fixed max. threading depth, you could make a “view” (Mongo pipeline) on the DB which translates ”my heavy model” into a “messageId, parentRoomId” scheme. From this view, a room’s total messages could be fetched and “colored” according to their thread. 
However, I fear you’d have to contribute it yourself. If you do, I can help you with the aggregation. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

mrsimpson commented:
@RocketChat/core I need some help:
It's about the "replies / last message age" indicators below the thread message in the parent channel.
There is no such thing as additional data to a message in the messages list yet. So I went for re-using the already existing attachment fields and add some custom rendering to it. idea: Additional data (such as a button which holds the rely counter) is a property of the message which can be presented differently. I had this on another PR already and picked it with efcad4b0d9f842b3b07f0dbe4921f0e4d3085d67 .
This allows me to get the button rendered without much hacking in the sensitive `rocketchat:lib` package.
However, I'm failing to make this reactive: The messages list is rendered non-reactively and it seems that no `onRendered` of a child template is called as a consequence.
I added some comments [here](https://github.com/assistify/Rocket.Chat/commit/548f77acaa381d5cc3b88bad5d6fbb70503aa7ef#diff-c1fce0b3cda38acd5f8028617b61c7cfR40) and [there](https://github.com/assistify/Rocket.Chat/commit/548f77acaa381d5cc3b88bad5d6fbb70503aa7ef#diff-c4fca4c8d9908a848651650e1706aee1R29) as a starting point to where I think something needs to be done.
Please help me out on how to overcome this pitfall, then I can do the rest.

rtarantelli commented:
Hi @RocketChatApp and @geekgonecrazy any news about this PR?

sampaiodiego commented:
thanks @rtarantelli for you interest =) as you can see I just fixed the conflicts. The idea is to give this a try and keep things going.

zdumitru commented:
@mrsimpson Is there anything you're missing now in order to proceed?

sampaiodiego commented:
@mrsimpson what about just using regular attachments fields and then update the message on each reply to the thread? so additional work would only be needed to do for the 'last reply', which would be a client side thing only, no need to get real time data from server.

btw, I have created https://threads-qa.rocket.chat to help everyone testing the feature. everyone is welcome to help testing it and providing feedback.

mrsimpson commented:
>@mrsimpson what about just using regular attachments fields and then update the message on each reply to the thread?...

@sampaiodiego this allows us to get rid of a subscription / streamer and I‘d consider it... ok, though race-conditions may occur. 
However, this does not solve my major issue: „how to make the rendered attachment field reactive“? Even if I update the field on the message attachment, the NRR of the messages list prevents a refresh of the UI, as far as I could debug. Or am I mistaken?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

sampaiodiego commented:
>„how to make the rendered attachment field reactive“? 

@mrsimpson if the message is updated it will re-render (that's how reaction works for example).

> though race-conditions may occur.

true. I would suggest to use [findAndModify](https://docs.mongodb.com/manual/reference/method/db.collection.findAndModify/) to avoid race-conditions but I read this:

> When modifying a single document, both db.collection.findAndModify() and the update() method atomically update the document.

so, we can try using different techniques, like using `$inc`.

sampaiodiego commented:
@mrsimpson both are tricky :disappointed: I'm not finding any solution already in place to make both work.

I was thinking on having this metadata on message's object, but after thinking on this solution I realized we already have them on room's object (properties `msgs` and `lm`), so adding the same data to another records doesn't seen right.

but now I'm conflicted on how that should be done.. for the threads the user is already joined, we have the rooms objects cached on the client, so we could use that. but for the others, idk what the most efficient way to do it.. need to think more about it.

mrsimpson commented:
As you probably already saw, I render the attachment fields with a dedicated Meteor template. I thought of utilizing `onCreated` for hooking up a listener. But I seems as if the NRR also blocks that. What about hooking up something like `Meteor.Streamer` as an alternative callback-channel? We could then explicitly trigger `renderWithData`, couldn’t we?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

sampaiodiego commented:
because of NRR nothing inside the message template will ever be re-rendered by any reactivity. the only way of doing it is updating something on the message object, that will trigger a re-render of the whole message component.

mrsimpson commented:
@sampaiodiego If there's nothing else triggering an update, then we'll have to have a redundant storage 👎 
I now implemented this in a race-condition-safe manner (I at least think so) and decoupled the update-process. Imho, it's as good as it can get now (given the NRR-boundaries).
However, during testing I noticed that the last-message age is also not getting re-rendered which is also necessary as time goes by (every second, the age changes without an update of the `lm`).
I can see two options:

- Don't display the age at all but only a counter
- Render the last message timestamp (instead of an age)

@geekgonecrazy I added the tabbar (copy-paste of pinned-messages as of now).

pshute commented:
@sampaiodiego creating https://threads-qa.rocket.chat for testing is a great idea.

Is it possible for me to open a thread someone else has created? I seem to only be able to open my own.

mrsimpson commented:
@zdumitru thx, merged `develop` fixing this conflict

mrsimpson commented:
@pshute 

> Is it possible for me to open a thread someone else has created? I seem to only be able to open my own.

It's a matter of permissions:
![image](https://user-images.githubusercontent.com/17176678/50795249-df921800-12cd-11e9-87ea-14e5a69d579e.png)

mrsimpson commented:
@sampaiodiego on the threads-qa-server, a major issue is the un-ability to resolve threads you're not an owner of. This is due to the bug that `FlowRouter.goToRoomById` uses `ChatSubscriptions` internally to generste the link. 
I fixed this in https://github.com/RocketChat/Rocket.Chat/pull/11745 long time back and separated it since this affects (fixes) all mentions. 
Can you please re-review this other PR and - if fine - merge into this one? I wanted to avoid that since it will enlarge this PR even more.

mrsimpson commented:
Documentation about the current state and what needs to be discussed / decided

## Obvious bugs

- [ ] navigation to the thread is broken if the user navigating is not subscribed to the thread. This is [actually solved](https://github.com/RocketChat/Rocket.Chat/pull/11803#issuecomment-452101924)

## Nice to haves

- The notification of new messages on the thread (by only updating the message count) is very subtle. Members of the thread will get the notifications as with any other room, but others may not notice them. However, any other highlighting (e. g. changing the color of the button) is more than tricky due to the non-reactivity: Changing the user's subscription would have to "touch" the linking message - which would be very hacky and a violation of the model.

## Conceptually unclear

- [ ] Who should be invited to the thread? Currently, all active users of a public channel will be selected since they're the ones to likely answer. For private groups, all members of the parent are dragged into the thread. This is necessary since there's no option up to now to actively join a private group. Without being able to request access to a room (see #11272), members of the parent thread would be able to see that a thread has been created, but they'd never be able to self-join the threaded discussion.
- [ ] Every room needs a unique name in RocketChat. This name is e. g. used for mentions and also for the general navigation (basically the URL). I don't see an option how this could change in near future. I discurage hiding the thread's name (e. g. by replacing it on the UI with the first message text), since there are many places in which the `name` will be used (Header, Directory, mentions, search, Sidepanel, ...)
Instead of creating the name via `parentRoomName+counter`, I can also imagine to use `parentRoomName+sluggifiedFirstMessage`, but a first look at this option resulted in unreadable names (`general-hello-all-what-is-the-next-best-option-to-dine` vs. `general-hello-all-what-is-the-next-best-option-to-improve-performance`). 

For both challenges (inviting users) and the name, I can very well imaging to present a form to the user on thread creation, prefilled with generate name and users so he can adapt that. But this adds much more complexity. For the user as well as for the one implementing this feature.

## Not even talked about

- [ ] Currently, there are a lot of texts being added to the thread explaining what has happened during threading. Unless you do this, the user who created the thread may sit alone in the room talking to himself without knowing how things shall continue. This is due to the heavy nature of the thread.
- [ ] How should the user be able to start a thread? (How) should a user be able to create a thread without being part of the channel? The current (big button in Sidepanel) is for sure not the beauty-award-winning-option.

With all the potential implementation options, we need to keep in mind that the (native) mobile apps have to at least be able to provide a similar experience. Even the rendering of the last message & thread message count is not available on mobile, thus no navigation to the thread is possible /cc @rafaelks 

Imho, the implementation as it currently is without major bugs (having applied the PR mentioned above). So I'm waiting for your, @RocketChat/core , particularly @sampaiodiego 's decision on how to continue.

robbyoconnor commented:
Can this be like slack? I fnd

robbyoconnor commented:
Threads should have the names of all participants and shouldn't show n the sidebar by default

robbyoconnor commented:
Another issue I disagree with is the fact that threads not their own entity, but rather rooms.

robbyoconnor commented:
You should probably also move all packages from the `assistify` namespace to `rocketchat` as well

pshute commented:
@robbyoconnor 
> Another issue I disagree with is the fact that threads not their own entity, but rather rooms.

I think it could work if it only displayed, say, the 5 threads with the most recent replies, with a button to show more.

mrsimpson commented:
>Can this be like slack?

@robbyoconnor well - it can. But this PR is - as stated above - implementing another approach. 
Model-wise, the methods are compatible (you can basically implement Slack-like threading visualization on top of the PR without needing another migration), but there are many tricky questions to sort out. 

Once you inline the thread, the way the user is notified about new messages in a thread needs to be completely redesigned. Also, the mobile apps need to support a similar mechanism. 

Since this PR is already quite heavy, I’ll definitely not implement it here. In fact adoption of PRs has been quite slow in the recent past, so I will not spend any more effort until I get feedback from @RocketChat:core 
Once it’s finalized, I will change the package name which is the smallest thing to do ;-)
 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

pshute commented:
Should we be expecting to see some changes to the test site at https://threads-qa.rocket.chat addressing some of the issues found?

mrsimpson commented:
@zegorax pointed out that users Ned to be part of the thread from the beginning and cannot join afterwards. There are two aspects of this statement:

- there is a bug in RC about Navigation to rooms one is not a member of. And there’s a fix I referred to above. It just needs to be added. 
- much more interesting imho is how Threads of private groups shall be considered with respect to authorization. Currently, all members of the group are therefore added to the thread. But also kn this case, someone joining the parent lateron has got no permission to join the private thread group. Bug or feature?

thiagosanchz commented:
Guys, we are keeping working to improve the usability of threads based on the solution proposed by @mrsimpson  After a lot of discussions, we improve three things that are essential to make threads easy and understandable to use.

**1.**
 All actions that start a thread will open a dialog that the user can edit the name of the thread or create a new one.  
![threads - message action](https://user-images.githubusercontent.com/39275608/52085780-7d56cb00-258c-11e9-987f-dc8086aa0eac.gif)

**2.**
On the first prototype made by @mrsimpson we can see a button on sidebar to create a new thread. But since thread is related on a channel, we design a way that user can create a new thread but inside this channel (on text area) and not on the sidebar. 
![threads - text area](https://user-images.githubusercontent.com/39275608/52086145-4634e980-258d-11e9-9757-0507174d24a0.gif)

**3**
We receive a lot of feedbacks about if the user has inactive threads, the sidebar will be full with inactive items. So, we create a button to hide all threads. 
![threads - hide all](https://user-images.githubusercontent.com/39275608/52086938-4d5cf700-258f-11e9-96f6-9e216a6cb525.gif)

Another point that we discuss a lot is if the user can understand the top bar. If they'll understand if the thread is related on the channel. We changed the title of the thread to the top and the channel name to below the thread name.

pshute commented:
@thiagosanchz 
>1. All actions that start a thread will open a dialog that the user can edit the name of the thread or create a new one.

The original thread-channel names were prefixed with the parent channel name. Will there be any way to tell which channel is the parent after a user has changed the thread-channel name?

pshute commented:
@thiagosanchz 
> 3 We receive a lot of feedbacks about if the user has inactive threads, the sidebar will be full with inactive items. So, we create a button to hide all threads.

This button will be helpful, but if I have many threads, I might be clicking that button all day to display and hide the thread channels. I think it would be helpful to have an option to show, say, just the five thread-channels with the most recent activity.

Will we be able to try these latest changes on the trial site soon?

mrsimpson commented:
@thiagosanchz appreciate every reaction to how things shall proceed!
However, I feel as if the decisions made are not the critical ones.

> **1.**
> All actions that start a thread will open a dialog that the user can edit the name of the thread or create a new one.

We had this on our fork - and user feedback was one click too much. Therefore, we removed the dialog again. However, with defaulting the users to be invited, this can provide meaningful additional value and / or explanation: Once you start a thread from a private group, all members of the parent have to be defaulted, since else they won't have access to the thread (given the current authorization concept).

> **2.**
> On the first prototype made by @mrsimpson we can see a button on sidebar to create a new thread. But since thread is related on a channel, we design a way that user can create a new thread but inside this channel (on text area) and not on the sidebar.

We introduced the button in the side panel, since else nobody noticed the new feature. It can be configured to be hidden, since for "experienced" teams, it might look a bit ugly.

With respect to the button inside the channel: It's already implemented, isn't it?
<img width="735" alt="bildschirmfoto 2019-02-01 um 08 42 54" src="https://user-images.githubusercontent.com/17176678/52109795-f7279c80-25fe-11e9-876c-a33822481afa.png">

btw: I also added a slash command, but the version deployed on `threads-qa` is very old

> **3.**
> We receive a lot of feedbacks about if the user has inactive threads, the sidebar will be full with inactive items. So, we create a button to hide all threads.
> ![threads - hide all](https://user-images.githubusercontent.com/39275608/52086938-4d5cf700-258f-11e9-96f6-9e216a6cb525.gif)

This is not only a problem related to threads: If users created semantical channels without threads in the past, the problem was also there for normal channels. 
I even created a PR https://github.com/RocketChat/Rocket.Chat/pull/12117 for that...

> We changed the title of the thread to the top and the channel name to below the thread name.

I like that. Where will the channel description of a thread go? Will is simply be concatenated with the parent channel?

Other than that: All my questions asked [about a month ago](https://github.com/RocketChat/Rocket.Chat/pull/11803#issuecomment-452229168) are still all unanswered. Is there any though about that as well?

Further more: Please let me know what you as maintainers want to implement and what you'd like me to contribute. I'm in general happy to, but I hate working ~for the bin~ never merged PRs.

pshute commented:
Reporting a bug found in the demo site at https://threads-qa.rocket.chat.

I started a thread from an existing message, then decided it would be better to reply to it. It appears the Reply option isn't available to messages that are part of a thread, so I got into the thread-channel I'd created, and deleted it.

In #general, the original message is still displayed as being part of a thread, but you can't enter it because it's not there anymore.

khaledalyawad commented:
I have created a forked branch with working threading 

https://github.com/cooderzinc/rocket.chat.reply.threading

![image](https://user-images.githubusercontent.com/7322671/52174928-bf386a80-2760-11e9-8737-c2ce80f55e06.png)

mrsimpson commented:
> I have created a forked branch with working threading

@khaledalyawad as far as I've understood from the screenshot, you visualize the thread in the tab bar, including the possibility to reply there. I hope that I'm right guessing that you did not change anything with respect to the backend.
If this is right, it matches my intention: Once you've got the model and mechanisms in place, you can implement different visualization options on top of it.

@khaledalyawad How did you manage reactivity?

khaledalyawad commented:
I just added a new property in the message Model called customFields, 
the parent message will have customFields.replyIds while the reply messages will have customFields.ref which is the reference id to the parent

you can complete this (because I forked from here) https://github.com/RocketChat/Rocket.Chat/commit/abe1f41202375641f2213afd29314e0fb6f22113

to this

https://github.com/cooderzinc/rocket.chat.reply.threading

mrsimpson commented:
> you can compare this (because I forked from here) [abe1f41](https://github.com/RocketChat/Rocket.Chat/commit/abe1f41202375641f2213afd29314e0fb6f22113)

Ah, then this is a completely alternative implementation of threading which actually has got nothing to do with this PR, right? (I'm not disregarding this, just want to know!)
If so, kindly open a new PR so that your code can be deployed, tested and discussed by the community more easily. Thanks!

khaledalyawad commented:
yes, I agree, I was sharing it here just to let you know that I have tackled the same problem with a different approach

zdumitru commented:
Milestone specified in this issue **0.74.0** is no longer applicable. :-(

MarcosSpessatto commented on packages/assistify-threading/server/models/MessagesExtension.js:
```diff
@@ -7,12 +7,12 @@ import { RocketChat } from 'meteor/rocketchat:lib';
  */
-Object.assign(RocketChat.models.Messages, {
+Object.assign(Messages, {
 	refreshThreadMetadata(linkMessageId) {
```
@ggazzo can we move these functions to inside `models` package?

MarcosSpessatto commented on packages/assistify-threading/server/models/RoomsExtension.js:
```diff
@@ -1,8 +1,8 @@
 import s from 'underscore.string';
 
-Object.assign(RocketChat.models.Rooms, {
+Object.assign(Rooms, {
```
@ggazzo can we move these functions to inside `models` package?

geekgonecrazy commented:
Alright!  It has landed!  Big huge kudos to @mrsimpson for initial work!

engelgabriel commented:
https://github.com/RocketChat/Rocket.Chat/issues/13843

engelgabriel commented:
> We have decided that was better to delay our v1 release and get the Threads done properly than get the feature out the way it was initially implemented on the develop branch. We have "rebranded" our current implementation of Threads to Discussions (was it creates a new room type) and will implement Threads exactly the same way as Slack does.
>
> In our vision there are 2 different use cases on how to break out of the main conversation: one very easy and organic, with a single click for small side talk, and other, much more structured, that creates a new conversation with its own life cycle.
>
> We have tried to have a single solution for both, and as a result, no one was happy with it.
So we have given up on having a single approach for both use cases and will implement both: the Slack-like industry standard and our brand new one, called Discussions.

https://github.com/RocketChat/Rocket.Chat/issues/13843
